Make `GraphEdit` toolbar more customizable

YuriSizov opened the pull request:
Depends on https://github.com/godotengine/godot/pull/81551 ~(drafting for now)~.
This should address the issues mentioned in https://github.com/godotengine/godot-proposals/issues/7061, possibly enough to close the proposal.

I'm introducing a set of exposed properties to hide parts of the menu bar, and the menu bar itself. Previously existing properties to hide the arrange button and to show the zoom label are grouped together with the new properties in the inspector. The property to hide the arrange button has been renamed, because its appearance in the inspector can be very misleading:

<img width="377" alt="godot windows editor dev x86_64_2023-09-12_19-36-03" src="https://github.com/godotengine/godot/assets/11782833/a47d8aa6-a953-4f70-adac-e55792dbc75d">

Now the property is called `hide_arrange_button`, and its methods are `set_hide_arrange_button` and `is_hiding_arrange_button`. Similar naming scheme was used for the new properties. The property to show the zoom label remains unchanged.

This can be discussed further because the problem of negative and positive names is a pretty annoying one. But for now this is my proposal.


https://github.com/godotengine/godot/assets/11782833/2d00b84f-2d2e-4443-80d4-db9756909eaf

----

Additionally I addressed a usability issue with the toolbar that I had noticed long ago. Since the toolbar didn't have a background and could easily appear on top of various text and icons, it could be hard to read and use under certain circumstances:

<img width="773" alt="godot windows editor dev x86_64_2023-09-12_19-07-41" src="https://github.com/godotengine/godot/assets/11782833/cf4a7f96-cfd5-40eb-b662-35860f6688d8">

So I added a semi-transparent background (both in the default theme and in the editor):

<img width="765" alt="godot windows editor dev x86_64_2023-09-12_19-28-03" src="https://github.com/godotengine/godot/assets/11782833/f0bc0e9a-37e9-49a9-a721-f24edbb9bfec">

<img width="576" alt="godot windows editor dev x86_64_2023-09-12_19-30-52" src="https://github.com/godotengine/godot/assets/11782833/ae86f2c3-0b0b-4a79-b954-cfcd8fd3f1b3">

PS. There are some changes in unrelated files because I've cleaned up includes in `graph_edit.h`.

lostdisplay commented:
Great stuff! Is it possible to give the user the option to hide the scrollbars too?

YuriSizov commented:
@lostdisplay Sure, but it would be a considerable usability downgrade to do so if panning and zooming is still allowed. What would be the use case?

lostdisplay commented:
@YuriSizov I can think of _a_ use case, and that's for my personal project 😅, an app for taking notes with an infinite whiteboard. The scroll bars would just be a hindrance. Right now I'm connecting the visibility_changed signal of the internal scrollbars to a function that hides them again, which is hacky at best. 

But to be honest, this has a very limited use case. I've seen some people using GraphEdits for skill trees though?

YuriSizov commented:
> I've seen some people using GraphEdits for skill trees though?

I haven't, and I'm not sure `GraphEdit` is that suitable for more simplistic game UIs. But I guess I've seen games where such UIs are done without any scroll indicators. I think it's a material for a different PR, but yeah, it should be doable, since it's `GraphEdit` that hides and shows them.

YuriSizov commented:
Rebased on `master` and also added compatibility methods. Properties cannot be added for compatibility at the moment. Something to look into in the future. GDExtensions don't have them, but C# might need us to do this gracefully. Can also be useful if we want to expose compatibility code to GDScript.

raulsntos commented:
> C# might need us to do this gracefully.

That's right, we would have to add the old properties to the [`Compat.cs`](https://github.com/godotengine/godot/blob/master/modules/mono/glue/GodotSharp/GodotSharp/Compat.cs) file if we want to avoid breaking compat.

That said, since `GraphEdit` is marked experimental, and we are already breaking compat in the `GraphNode` API, is there any point to providing these compat methods/properties?

YuriSizov commented:
> That said, since `GraphEdit` is marked experimental, and we are already breaking compat in the `GraphNode` API, is there any point to providing these compat methods/properties?

Outside of the automated system that we have, no, I don't think so. But for GDExt we want to keep binary compatibility as intact as possible, so methods at least are worth adding when we remove or rename them.

raulsntos commented:
I don't disagree with that. My point was that if we are already breaking binary compatibility in the `GraphNode` API, adding these compat methods won't help those users, right? They would still be affected by the break.

YuriSizov commented:
> My point was that if we are already breaking binary compatibility in the `GraphNode` API, adding these compat methods won't help those users, right? They would still be affected by the break.

Most of it was renames, so a compatibility method that routes to the new method should be sufficient for GDExtension users who may be using them. That said, I'm not particularly sure where the line would be in this specific case.

Geometror commented on scene/gui/graph_edit.cpp:
```diff
@@ -1878,15 +1935,19 @@ void GraphEdit::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::VECTOR2, "minimap_size", PROPERTY_HINT_NONE, "suffix:px"), "set_minimap_size", "get_minimap_size");
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "minimap_opacity"), "set_minimap_opacity", "get_minimap_opacity");
 
 	ADD_GROUP("UI", "");
```
I think "UI" is a bit too general here. What about "Menu" or "Toolbar"?

YuriSizov replied:
Ah, what the hell, let's make it both! "Toolbar Menu" :)

Geometror commented on scene/gui/graph_edit.cpp:
```diff
@@ -1878,15 +1935,19 @@ void GraphEdit::_bind_methods() {
 	ADD_GROUP("UI", "");
-	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "arrange_nodes_button_hidden"), "set_arrange_nodes_button_hidden", "is_arrange_nodes_button_hidden");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "hide_menu"), "set_hide_menu", "is_hiding_menu");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "show_zoom_label"), "set_show_zoom_label", "is_showing_zoom_label");
```
IMO these properties should either be all negative or all positive. With the compatibility methods this shouldn't be a problem (introducing methods for show_zoom_label [all negative] or change the ones you already wrote [all positive]).
Beyond that I have no opinion on which (all positive or all negative) is better.

YuriSizov replied:
I understand the appeal, and I would support it personally. But there was an alternative POV that checkboxes should be off by default, and any alternative behavior that you opt into should be enabled by toggling them. So because preserving the name would be slightly more compatibility friendly, I decided to go in the direction of that alternative POV.

I also considered that switching the polarity is a more annoying compat change than a simple rename. You can't just search and replace it, you have to invert the logic everywhere. So for better or worse, I'd stick to the white raven `show_zoom_label` being what it is.

But ultimately, I'm fine either way, as long as we change `arrange_nodes_button_hidden` 🙃

akien-mga replied:
I was arguing for off-by-default checkboxes, but to be honest I'm not that convinced that it's the absolute best design. I think here a `show_menu` that's enabled by default would also be ok.

So pick what you prefer, and it should be fine :)

YuriSizov replied:
Alright, the I guess it would make more sense to invert `arrange_nodes_button_hidden` and keep every property as `show_*` with some being on and some being off by default.

Geometror reviewed (commented):
Overall solid work, the visual usability improvement looks nice!
I was actually just planning to implement this after I took another look at my GraphEdit refactoring todo list. Luckily I noticed your PR before that :)

YuriSizov commented:
Changed all toolbar properties to be `show_*` (with `show_zoom_label` being the only one off by default, which is its current state).

<img width="190" alt="godot windows editor dev x86_64_2023-09-26_13-21-53" src="https://github.com/godotengine/godot/assets/11782833/28d32dab-c8c0-4b33-aea2-b5031886a1da">

With the flat button changes the toolbar currently looks like this

<img width="576" alt="godot windows editor dev x86_64_2023-09-26_13-21-36" src="https://github.com/godotengine/godot/assets/11782833/94fae091-202d-48da-be82-52028b9398ab">
<img width="922" alt="godot windows editor dev x86_64_2023-09-26_13-20-56" src="https://github.com/godotengine/godot/assets/11782833/89beda64-0513-448d-8010-9d55f4135e62">

elmanjar13 commented:
Thats' good!!!!

YuriSizov committed:
Make `GraphEdit` toolbar more customizable

Also adds a semi-transparent background
to make the toolbar stand out in front of nodes.

akien-mga commented:
Thanks!
